src/rust/lib/graph-editor/src/component/node/port/output.rs

```diff
@@ -0,0 +1,682 @@
+}
+
+impl Frp {
+    fn new(network: &frp::Network) -> Self {
+        frp::extend! { network
+            def set_size           = source();
+            def on_port_mouse_down = source();
+
+            let port_mouse_down = on_port_mouse_down.clone_ref().into();
+        }
+        Self{set_size,port_mouse_down,on_port_mouse_down}
+    }
+}
+
+
+
+// =======================
+// === OutputPortsData ===
+// =======================
+
+/// Internal data of the `OutputPorts`.
+#[derive(Debug)]
+pub struct OutputPortsData {
+    display_object : display::object::Instance,
+    logger         : Logger,
+    size           : Cell<Vector2<f32>>,
+    gap_width      : Cell<f32>,
+    ports          : RefCell<ShapeView>,
+}
+
+impl OutputPortsData {
+
+    fn new(scene:Scene, number_of_ports:NonZeroU32) -> Self {
+        let logger         = Logger::new("OutPutPorts");
+        let display_object = display::object::Instance::new(&logger);
+        let size           = Cell::new(Vector2::zero());
+        let gap_width      = Cell::new(SEGMENT_GAP_WIDTH);
+        let ports          = ShapeView::new(number_of_ports, &logger, &scene);
+        let ports          = RefCell::new(ports);
+
+        OutputPortsData {display_object,logger,size,ports,gap_width}.init()
+    }
+
+    fn init(self) -> Self {
+        self.ports.borrow().set_parent(&self.display_object);
+        self.update_shape_layout_based_on_size_and_gap();
+        self
+    }
+
+    fn update_shape_layout_based_on_size_and_gap(&self) {
+        let size      = self.size.get();
+        let gap_width = self.gap_width.get();
+        self.ports.borrow().update_shape_layout_based_on_size_and_gap(size, gap_width);
+    }
+
+    fn set_size(&self, size:Vector2<f32>) {
+        self.size.set(size);
+        self.update_shape_layout_based_on_size_and_gap();
+    }
+}
+
+
+
+// ===================
+// === OutputPorts ===
+// ===================
+
+/// Implements the segmented output port area. Provides shapes that can be attached to a `Node` to
+/// add an interactive area with output ports.
+///
+/// The `OutputPorts` facilitate the falling behaviour:
+///  * when one of the output ports is hovered, after a set time, all ports are show and the hovered
+///    port is highlighted.
+///  * when a different port is hovered, it is highlighted immediately.
+///  * when none of the ports is hovered all of the `OutputPorts` disappear. Note: there is a very
+///    small delay for disappearing to allow for smooth switching between ports.
+///
+#[derive(Debug,Clone,CloneRef)]
+pub struct OutputPorts {
+    /// The FRP api of the `OutPutPorts`.
+    pub frp     : Frp,
+        network : frp::Network,
+        data    : Rc<OutputPortsData>,
+}
+
+impl OutputPorts {
+    /// Constructor.
+    pub fn new(scene:&Scene, number_of_ports:NonZeroU32) -> Self {
+        let network = default();
+        let frp     = Frp::new(&network);
+        let data    = OutputPortsData::new(scene.clone_ref(), number_of_ports);
```

Comment:
Lets not design an API which requires cloning on usage side. Pass `&scene` instead. If you want to make it generic, you can always use `scene: impl Into<Scene>` because there is `impl From<&Scene> for Scene` defined.